### PR TITLE
[CALCITE-6300] Function MAP_VALUES/MAP_KEYS gives exception when mapVauleType and mapKeyType not equals map Biggest mapKeytype or mapValueType

### PR DESCRIPTION
**Pull request by caicancai**
https://issues.apache.org/jira/browse/CALCITE-6300

### CODE REVIEW

**Comment by caicancai** [comment]
https://issues.apache.org/jira/browse/CALCITE-6300

**Comment by caicancai** [comment]
I think I can encapsulate a unified interface during parameter verification to convert all spark map functions

**Comment by caicancai** [comment]
@mihaibudiu Hello, I converted the callbinding after parameter verification, which only applies to the map function of spark.
I debugged the spark source code and found that the type has been converted before entering the logical plan.

What do you think

File: core/src/main/java/org/apache/calcite/sql/type/NonNullableAccessors.java [hunk 1]
```diff
@@ -58,4 +58,10 @@ public static RelDataType getKeyTypeOrThrow(RelDataType type) {
     return requireNonNull(type.getKeyType(),
         () -> "keyType is null for " + type);
   }
+
+  @API(since = "1.37", status = API.Status.EXPERIMENTAL)
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/type/NonNullableAccessors.java:62]: Is this annotation necessary?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 2]
```diff
@@ -8182,6 +8182,19 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
         "CHAR(3) NOT NULL ARRAY NOT NULL");
     f1.checkScalar("map_keys(map('foo', 1, null, 2))", "[foo, null]",
         "CHAR(3) ARRAY NOT NULL");
+
+    f1.checkScalar("map_keys(map('foo', 1, null, 2))", "[foo, null]",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:8186]: I am assuming these have been checked against spark.
Perhaps you can add a comment, also listing the spark version - Spark has been known to change function behaviors in the past.

File: core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java [hunk 3]
```diff
@@ -1491,6 +1496,43 @@ private static class PeriodOperandTypeChecker
...
+      }
+
+      final RelDataType mapKeyType =
+          getKeyTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0));
+      final RelDataType mapValueType =
+          getValueTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0));
+
+      // If the key and value types are not unknown,
+      // adjust their types to suit the Map function's constructor.
+      if (mapKeyType.getSqlTypeName() != SqlTypeName.UNKNOWN
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:1528]: I think we have discussed this in the past: I don't think that the type checker is supposed to modify the node that is being checked.

There are type checking strategies that will backtrack (e.g., SqlOperandTypeChecker.or), and they do not expect the checked data structure to change.

I think this pattern is very dangerous, and can lead to hard to diagnose bugs.

On the other hand, I don't know how this should be done properly. The map function does need to introduce type casts. But maybe the typechecking stage is not the right stage.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 2 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: done, I tested all versions after spark3.4

File: core/src/main/java/org/apache/calcite/sql/type/NonNullableAccessors.java (hunk 1 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/type/NonNullableAccessors.java]: done

File: core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java (hunk 3 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:1528]: Currently, this type check is specially prepared for Spark's map function, but there may be some problems that are difficult to eliminate as you said. However, Spark currently performs some type conversions during parameter checking. I have to admit that Spark does not do this very well. 
 
I believe that after I have a deeper understanding of Calcite and the type system in the future, I can improve this part. I hope that day will not be too far away.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:1528]: As Julian said in the Jira comments, you should insert casts during validation, before type checking.

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:1528]: ok, thank you.

File: core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java [hunk 4]
```diff
@@ -1532,6 +1532,7 @@ private static class MapKeyOperandTypeChecker extends SameOperandTypeChecker {
     }
   }
 
+
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:1535]: This new line seems an unnecessary change, please undo.

**Comment by rubenada** [comment]
@mihaibudiu do you have further comments here? IMO the proposed solution seems a bit ad-hoc, but seems to do the trick (and the new code is very well located). Shall we consider this for 1.38? (after resolving conflicts, squashing commits, etc.)

**Comment by caicancai** [comment]
@rubenada Looks like my local git crashed for me, I will open a new PR
